sql/parser: encodeSQLIdent double quotes lookahead keywords

eisenstatdavid opened the pull request:
encodeSQLIdent currently does not quote non-reserved keywords, which
causes statements like
`SELECT * FROM t AS "of" AS OF SYSTEM TIME '2016-01-01'` to be formatted
improperly.

Since encodeSQLIdent is the only user of reservedKeywords, delete it and
its generating script.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/15822)
<!-- Reviewable:end -->

eisenstatdavid commented:
I meant keywords, not identifiers (that would be silly). Fixed the commit message.

---

Review status: 0 of 6 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15822#-:-Kjixlhwm1cSnnF06g9C:bddw1wq)*
<!-- Sent from Reviewable.io -->

mjibson commented:
Review status: 0 of 6 files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

*[pkg/sql/parser/scan.go, line 735 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15822#-KjiyhZE0XhjxeiqapQ8:-KjiyhZE0XhjxeiqapQ9:btz6vmh) ([raw file](https://github.com/cockroachdb/cockroach/blob/7655e9305d2d4dc811d73293e5672fcdacb7475d/pkg/sql/parser/scan.go#L735)):*
> ```Go
> 		}
> 	}
> 	_, ok := reservedKeywords[strings.ToUpper(s)]
> ```

I don't understand why this was removed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15822)*
<!-- Sent from Reviewable.io -->

eisenstatdavid commented:
Review status: 0 of 6 files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

*[pkg/sql/parser/scan.go, line 735 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15822#-KjiyhZE0XhjxeiqapQ8:-Kjiz5kfVz9Dpbh3bM3b:bijz6o6) ([raw file](https://github.com/cockroachdb/cockroach/blob/7655e9305d2d4dc811d73293e5672fcdacb7475d/pkg/sql/parser/scan.go#L735)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary><blockquote>

I don't understand why this was removed.
</blockquote></details>

`isIdent` has only one caller, `encodeSQLIdent`, so I moved the check there. `isIdent` is a bit of misnomer, since any string is a valid identifier if quoted. I'll reorganize and ping you.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15822)*
<!-- Sent from Reviewable.io -->

eisenstatdavid commented:
Ready for another look.

---

Review status: 0 of 6 files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15822#-:-Kjj-xWbL1sNQo8WOkpi:b467oqh)*
<!-- Sent from Reviewable.io -->

mjibson commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 6 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15822#-:-KjkMyWRvDfFY9kQqmck:bnfp4nl)*
<!-- Sent from Reviewable.io -->

eisenstatdavid commented:
The tests are broken, seemingly because some uses of `Name` are not coping with the new quotes. I'm going to explore the trade-offs between fixing these, special casing the potentially `_LA` tokens, and programmatic derivation from the parse tables of "problematic" identifiers that is less than the reserved set.

Commit:
sql/parser: encodeSQLIdent double quotes lookahead keywords

encodeSQLIdent currently does not quote non-reserved keywords, which
causes statements like
`SELECT * FROM t AS "of" AS OF SYSTEM TIME '2016-01-01'` to be formatted
improperly.

eisenstatdavid commented:
Scaled this PR back to quote only the tokens that cause an _LA token. I have a hunch that that's not enough, but if so, I'll defer that work to a subsequent PR.

---

Review status: 0 of 7 files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15822#-:-Kjn9NGYkpEkNYgliaWf:b-3otfls)*
<!-- Sent from Reviewable.io -->

eisenstatdavid commented:
(PTAL)

mjibson commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 7 files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15822#-:-KjnP996BeVbxsxE6cIP:bnfp4nl)*
<!-- Sent from Reviewable.io -->
